contributors/design-proposals/vertical-pod-autoscaler.md

```diff
@@ -0,0 +1,486 @@
+and reducing the maintenance cost, by automating configuration of resource
+requirements. While ensuring high utilization it tries to minimize the risk of
+containers running out of memory or getting CPU starved.
+
+### Related features ###
+#### Initial resources ####
+["Initial Resources"](https://github.com/kgrygiel/community/blob/master/contributors/design-proposals/initial-resources.md)
+is a very preliminary, proof-of-concept feature providing initial request based
+on historical utilization. It is designed to only kick in on Pod creation.
+VPA is intended to supersede this feature.
+
+#### In-place updates ####
+In-place Pod updates is a planned feature to allow changing resources
+(request/limit) of existing containers without killing them, assuming sufficient
+free resources available on the node. Vertical Pod Autoscaler will greatly
+benefit from this ability, however it is not considered a blocker for the MVP.
+#### Resource estimation ####
+Resource estimation is another planned feature, meant to improve node resource
+utilization by temporarily reclaiming unused resources of running containers.
+It is different from Vertical Autoscaling in that it operates on a shorter
+timeframe (using only local, short-term history), re-offers resources at a
+lower quality, and does not provide initial resource predictions.
+VPA and resource estimation are complementary. Details will follow once
+Resource Estimation is designed.
+
+------------
+Requirements
+------------
+
+### Functional ###
+
+1. VPA is capable of setting container resources (CPU & memory request/limit) at
+   Pod submission time.
+   
+2. VPA is capable of adjusting container resources of existing Pods, in
+   particular reacting to CPU starvation and container OOM events.
+   
+3. When VPA restarts Pods it respects the disruption budget.
+
+4. It is possible for the user to configure VPA with fixed constraints on
+   resources, specifically: min & max request.
+   
+5. VPA is compatible with Pod controllers, at least with Deployments.
+   In particular:
+     * Updates of resources do not interfere/conflict with spec updates.
+     * It is possible to do a rolling update of the VPA policy on an existing
+       Deployment.
+     
+6. It is possible to create Pod(s) that start following the VPA  policy
+   immediately. In particular such Pods must not be scheduled until VPA policy
+   is applied.
+   
+7. Disabling VPA is easy and fast ("panic button"), without disrupting existing
+   Pods.
+
+### Availability ###
+1. Downtime of heavy-weight components (database/recommender) must not block
+   recreating existing Pods. Components on critical path for Pod creation
+   (admission controller) are designed to be highly available.
+
+### Extensibility ###
+1. VPA is capable of performing in-place updates once they become available.
+
+------
+Design
+------
+
+### Overview ###
+(see further sections for details and justification)
+
+1. We introduce a new type of **API resource**:
```

Comment:
How would we prevent limit range from fighting with VPA?  Does VPA have to be earlier in the admission chain than the LimitRange?